app/api/utils/AppContext.ts

```diff
@@ -0,0 +1,33 @@
+import { AsyncLocalStorage } from 'async_hooks';
+
+interface ContextData {
+  [k: string]: unknown;
+}
+
+class AppContext {
+  private storage = new AsyncLocalStorage<ContextData>();
+
+  async run(cb: () => Promise<void>, data: ContextData = {}): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.storage.run(data, () => {
+        cb()
+          .then(resolve)
+          .catch(reject);
+      });
+    });
+  }
+
+  get(key: string) {
+    const data = this.storage.getStore();
+    return data && data[key];
```

Comment:
@fnocetti i think we should manage this condition ?
in the original implementation we were throwing an error:
https://github.com/huridocs/uwazi/blob/3ed63fcd70dedf59bbbcc11042ed4b20f71dc6e1/app/api/tenants/tenantContext.ts#L59-L61

the reason for this is  because there was a bug where on certain cases the async context was lost, this way we have detected several processes we had more easily, where this was the case([reference](https://github.com/huridocs/uwazi/issues/2393#issuecomment-627775969)). 
I think that implementation belongs here now ? (apparently this bug was fixed on node but unless we know all the details on scenarios where the context can be lost, i think its safe to have it)

it is also a non tested condition.

Reply:
So happy you brought this topic! So, the easiest scenario in which you could have your data undefined is when you are running get or set outside of a context. That's basically what (I understood) the async hooks do.
My opinion is that we should not allow the use of the context variables outside of a context scope and throw an exception, but I didn't want to limit the functionality before having this discussion.
That said, I will do both:

- Keep the checking in the tenant side (the context might exist but be missing tenant information because of a tenant bug)
- Throw an exception if illegally accessing the getter/setter outside of a context scope (covering also possible bugs on node's side)

I'll also improve the testing coverage.

Thanks!

Reply:
sounds good to me 👍